lib/ansible/modules/cloud/amazon/ec2_group.py

```diff
@@ -840,147 +874,68 @@ def main():
-                                             ('0.0.0.0/0',
-                                              group['GroupName'],
-                                              e),
-                                         exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
-                changed = True
+            rule = {'IpProtocol': '-1', 'IpRanges': [{'CidrIp': '0.0.0.0/0'}]}
+            rule_id = make_rule_key('out', rule, group['GroupId'], '0.0.0.0/0')
+            if rule_id not in current_egress:
+                new_egress_permissions.append(rule)
             else:
-                # make sure the default egress rule is not removed
-                del groupRules[default_egress_rule]
-
-        # Finally, remove anything left in the groupRules -- these will be defunct rules
-        if purge_rules_egress and 'VpcId' in group:
-            for (rule, grant) in groupRules.values():
-                # we shouldn't be revoking 0.0.0.0 egress
-                if grant != '0.0.0.0/0':
-                    ip_permission = serialize_revoke(grant, rule)
-                    if not module.check_mode:
-                        try:
-                            client.revoke_security_group_egress(GroupId=group['GroupId'], IpPermissions=[ip_permission])
-                        except botocore.exceptions.ClientError as e:
-                            module.fail_json(msg="Unable to revoke egress for ip %s security group '%s' - %s" %
-                                                 (grant, group['GroupName'], e),
-                                             exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
-                    changed = True
+                present_egress.append(Rule(rule=rule, rule_id=rule_id, ip_permission=rule, ipv4='0.0.0.0/0', ipv6=None))
+
+        # Find rules to purge
+        purge_ingress_rule_ids = []
+        purge_egress_rule_ids = []
+        if purge_rules:
+            purge_ingress_rule_ids = [rule_id for rule_id in current_ingress if rule_id not in user_ingress_rule_ids]
+        if purge_rules_egress:
+            for rule_id in current_egress:
+                if rule_id not in user_egress_rule_ids:
+                    if rules_egress:
+                        purge_egress_rule_ids.append(rule_id)
+                    else:
+                        if not rule_id.endswith('0.0.0.0/0'):
+                            purge_egress_rule_ids.append(rule_id)
+
+        # Assemble ip permissions to remove
+        revoke_ingress, revoke_egress = assemble_permissions_to_revoke(purge_ingress_rule_ids, purge_egress_rule_ids,
+                                                                       current_ingress, current_egress)
+
+        # Revoke old rules
+        changed = remove_old_permissions(client, module, revoke_ingress, revoke_egress, group['GroupId'], changed)
```

Comment:
Instead of passing `changed` in on all these, can you make these assignments all `changed |= [method](.....)`?